opm/input/eclipse/EclipseState/Aquifer/NumericalAquifer/SingleNumericalAquifer.cpp

```diff
@@ -64,6 +69,33 @@ namespace Opm {
         return this->id_;
     }
 
+    void SingleNumericalAquifer::applyMinPV(const EclipseGrid& grid) {
+        constexpr auto DEFAULT_MINPV = 1.0e-6;
+        constexpr auto EPSILON = std::numeric_limits<double>::epsilon();
+        const auto& minpv_vector = grid.getMinpvVector();
+        const bool minpv_active = (grid.getMinpvMode() != MinpvMode::Inactive);
+        std::vector<std::size_t> invalid_cell_indices;
+        for (auto& cell : this->cells_) {
+            const double minpv = minpv_active ? minpv_vector[cell.global_index] : DEFAULT_MINPV;
+            if (cell.poreVolume() < minpv) {
+                if (const auto bulk_vol = cell.cellVolume(); bulk_vol > EPSILON) { // Scale porosity to reach MINPV, if possible
+                    cell.porosity = minpv/bulk_vol;
+                    const auto[I, J, K] = grid.getIJK(cell.global_index);
+                    OpmLog::warning(fmt::format("Pore volume in numerical aquifer {} cell ({}, {}, {}) below threshold - reset to MINPV = {} by adjusting PORO to {})",
+                                        this->id_, I + 1, J + 1, K + 1, minpv, cell.porosity));
+                } else { // Otherwise remove cell (done below)
+                    invalid_cell_indices.push_back(cell.global_index);
+                }
+            }
+        }
+        for (const auto cell_global_index : invalid_cell_indices) { // Inefficient, but also infrequent, so no issue in practice (?)
```

Comment:
For the cells that we need to remove based on `epsilon`, we might just not add them to `this->cells_`.  then the function `applyMinPV` will be simpler with only one purpose to scale up the cells with pore volume smaller than MinPV. 

And also, if we remove cells, we might need to check whether all the aquifers have at least one aquifer cell. (not sure whether we need to do something though, might work directly as if there is no aquifer existing)

It is just comment for discussion.

Reply:
Thanks, very good points! Tried to update accordingly now. Tested on a numaq-case from opm-tests, seemed to work fine (both scaling to minpv and removing too low bulk volume cells, including removing all of them.)